pep-0646.rst

```diff
@@ -414,6 +355,100 @@ Normal ``TypeVar`` instances can also be prefixed and/or suffixed:
     z = prefix_tuple(x=0, y=(True, 'a'))
     # Inferred type of z is Tuple[int, bool, str]
 
+Unpacking a Tuple Type
+----------------------
+
+We mentioned that a ``TypeVarTuple`` stands for a tuple of types.
+Since we can unpack an ``TypeVarTuple``, for consistency, we also
+allow unpacking a tuple type. As we shall see, this also enables a
+number of interesting features.
+
+
+Unpacking a Concrete Tuple Type
+'''''''''''''''''''''''''''''''
+
+Unpacking a concrete tuple type is analogous to unpacking a tuple of
+values at runtime. ``Tuple[int, *Tuple[bool, bool], str]`` is
+equivalent to ``Tuple[int, bool, bool, str]``.
+
+Unpacking an Unbounded Tuple Type
+'''''''''''''''''''''''''''''''''
+
+Unpacking an unbounded tuple preserves the unbounded tuple as it is.
+That is, ``*Tuple[int, ...]`` remains ``*Tuple[int, ...]``; there's no
+simpler form. This enables us to specify types such as ``Tuple[int,
+*Tuple[str, ...], str]`` - a tuple type where the first element is
+guaranteed to be of type ``int``, the last element is guaranteed to be
+of type ``str``, and the elements in the middle are zero or more
+elements of type ``str``. Note that ``Tuple[*Tuple[int, ...]]`` is
+equivalent to ``Tuple[int, ...]``.
+
+Unpacking unbounded tuples is also useful in function signatures where
+we don't care about the exact elements and don't want to define an
+unnecessary ``TypeVarTuple``:
+
+::
+
+    def process_batch_channels(
+        x: Array[Batch, *Tuple[Any, ...], Channels]
+    ) -> None:
+        ...
+
+
+    x: Array[Batch, Height, Width, Channels]
+    process_batch_channels(x)  # OK
+    y: Array[Batch, Channels]
+    process_batch_channels(y)  # OK
+    z: Array[Batch]
+    process_batch_channels(z)  # Error: Expected Channels.
+
+
+We can also pass a ``*Tuple[int, ...]`` wherever a ``*Ts`` is
+expected. This is useful when we have particularly dynamic code and
+cannot state the precise number of dimensions or the precise types for
+each of the dimensions. In those cases, we can smoothly fall back to
+an unbounded tuple:
+
+::
+
+    y: Array[*Tuple[Any, ...]] = read_from_file()
+
+    def expect_variadic_array(
+        x: Array[Batch, *Shape]
+    ) -> None: ...
+
+    expect_variadic_array(y)  # OK
```

Comment:
Why is this okay? y's shape may have length 0 (the end case of `Tuple[Any, ...]`) but x needs to have at least one dimension (`Batch`).

Reply:
`Tuple[X, ...]` represents a tuple of arbitrary length, not zero. For example, `xs: Tuple[int, ...]; xs[0]` is not a type error. It can be unsafe, yes, but that comes with the territory when we use `Tuple[int, ...]`.

This is what enables gradual typing of a large code base. A variable that hasn't yet been given shape types (`x: Array`) can be used without complaint wherever a precise shape type is expected (`y: Array[A, B, C]` or `y: Array[A, *Ts]`). I had added this example and the paragraph below to demonstrate this.